### PR TITLE
Graceful exit from CLI even after invest exception

### LINKED ISSUE #580: CLI - graceful exit after an invest model error
This originally came up as https://github.com/natcap/invest-workbench/issues/131 where anytime a model's `execute` errors, the traceback is sent to stderr, but then the cli also emits `'Failed to execute script cli'` to stderr. That message is associated with pyinstaller-made executables. 

We can avoid that message by always being sure to exit gracefully from `cli.py`. We do that elsewhere in the module, but aren't doing it right now for the `invest run --headless` command. I think we should avoid the message in this case because it's not relevant to the invest model's exception, and `cli.py` did not really fail to execute in this case.

### PR DESCRIPTION
**Pull request by davemfish**
This fixes #580 . Aside from preventing the trailing 'Failed to execute script cli` message, it also changes the traceback slightly as it now doesn't go all the way back to include 
```
  File "natcap\invest\cli.py", line 663, in <module>
  File "natcap\invest\cli.py", line 590, in main
  File "natcap\invest\carbon.py", line 317, in execute
```
That makes sense, but I can't really explain why there's a different number of Traceback blocks after this change. Maybe there's a different/better way to fix #580?

See the full before & after tracebacks:

**Before:**
```
07/09/2021 11:05:32  pygeoprocessing.geoprocessing ERROR    exception encountered in raster_calculator
Traceback (most recent call last):
  File "pygeoprocessing\geoprocessing.py", line 440, in raster_calculator
  File "pygeoprocessing\geoprocessing.py", line 1849, in _map_dataset_to_value_op
pygeoprocessing.geoprocessing.ReclassificationMissingValuesError: ('The following 11 raster values [52 53 54 55 56 57 58 59 60 61 62] from "C:\\Users\\dmf\\projects\\invest\\data\\bad-data\\Carbon\\lulc_current_willamette.tif" do not have corresponding entries in the ``value_map``: {1: 1.35, 2: 0.45, 3: 0.18, 4: 0.0, 5: 0.9, 6: 0.0, 7: 0.0, 8: 0.0, 9: 0.0, 10: 0.0, 11: 0.0, 16: 0.0, 102: 0.0}.', array([52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62], dtype=int16))
07/09/2021 11:05:32  pygeoprocessing.geoprocessing INFO     Waiting for raster stats worker result.
07/09/2021 11:05:32  pygeoprocessing.geoprocessing_core WARNING  No valid pixels were received, sending None.
07/09/2021 11:05:32  taskgraph.Task     ERROR    Something went wrong when adding task carbon_map_c_above_cur (0), terminating taskgraph.
Traceback (most recent call last):
  File "natcap\invest\utils.py", line 863, in reclassify_raster
  File "pygeoprocessing\geoprocessing.py", line 1856, in reclassify_raster
  File "pygeoprocessing\geoprocessing.py", line 440, in raster_calculator
  File "pygeoprocessing\geoprocessing.py", line 1849, in _map_dataset_to_value_op
pygeoprocessing.geoprocessing.ReclassificationMissingValuesError: ('The following 11 raster values [52 53 54 55 56 57 58 59 60 61 62] from "C:\\Users\\dmf\\projects\\invest\\data\\bad-data\\Carbon\\lulc_current_willamette.tif" do not have corresponding entries in the ``value_map``: {1: 1.35, 2: 0.45, 3: 0.18, 4: 0.0, 5: 0.9, 6: 0.0, 7: 0.0, 8: 0.0, 9: 0.0, 10: 0.0, 11: 0.0, 16: 0.0, 102: 0.0}.', array([52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62], dtype=int16))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "taskgraph\Task.py", line 697, in add_task
  File "taskgraph\Task.py", line 1174, in _call
  File "natcap\invest\carbon.py", line 442, in _generate_carbon_map
  File "natcap\invest\utils.py", line 871, in reclassify_raster
ValueError: Values in the LULC raster were found that are not represented under the 'lucode' column of the Carbon Pools table. The missing values found in the LULC raster but not the table are: [52 53 54 55 56 57 58 59 60 61 62].
07/09/2021 11:05:32  natcap.invest.utils INFO     Elapsed time: 0.17s
Traceback (most recent call last):
  File "natcap\invest\utils.py", line 863, in reclassify_raster
  File "pygeoprocessing\geoprocessing.py", line 1856, in reclassify_raster
  File "pygeoprocessing\geoprocessing.py", line 440, in raster_calculator
  File "pygeoprocessing\geoprocessing.py", line 1849, in _map_dataset_to_value_op
pygeoprocessing.geoprocessing.ReclassificationMissingValuesError: ('The following 11 raster values [52 53 54 55 56 57 58 59 60 61 62] from "C:\\Users\\dmf\\projects\\invest\\data\\bad-data\\Carbon\\lulc_current_willamette.tif" do not have corresponding entries in the ``value_map``: {1: 1.35, 2: 0.45, 3: 0.18, 4: 0.0, 5: 0.9, 6: 0.0, 7: 0.0, 8: 0.0, 9: 0.0, 10: 0.0, 11: 0.0, 16: 0.0, 102: 0.0}.', array([52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62], dtype=int16))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "natcap\invest\cli.py", line 663, in <module>
  File "natcap\invest\cli.py", line 590, in main
  File "natcap\invest\carbon.py", line 317, in execute
  File "taskgraph\Task.py", line 697, in add_task
  File "taskgraph\Task.py", line 1174, in _call
  File "natcap\invest\carbon.py", line 442, in _generate_carbon_map
  File "natcap\invest\utils.py", line 871, in reclassify_raster
ValueError: Values in the LULC raster were found that are not represented under the 'lucode' column of the Carbon Pools table. The missing values found in the LULC raster but not the table are: [52 53 54 55 56 57 58 59 60 61 62].
[11836] Failed to execute script cli
```

**After:**
```
2021-07-09 10:56:41,284 (pygeoprocessing.geoprocessing) geoprocessing.raster_calculator(495) ERROR exception encountered in raster_calculator
Traceback (most recent call last):
  File "pygeoprocessing\geoprocessing.py", line 440, in raster_calculator
  File "pygeoprocessing\geoprocessing.py", line 1849, in _map_dataset_to_value_op
pygeoprocessing.geoprocessing.ReclassificationMissingValuesError: ('The following 11 raster values [52 53 54 55 56 57 58 59 60 61 62] from "C:\\Users\\dmf\\projects\\invest\\data\\bad-data\\Carbon\\lulc_current_willamette.tif" do not have corresponding entries in the ``value_map``: {1: 1.35, 2: 0.45, 3: 0.18, 4: 0.0, 5: 0.9, 6: 0.0, 7: 0.0, 8: 0.0, 9: 0.0, 10: 0.0, 11: 0.0, 16: 0.0, 102: 0.0}.', array([52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62], dtype=int16))
2021-07-09 10:56:41,288 (pygeoprocessing.geoprocessing) geoprocessing.raster_calculator(510) INFO Waiting for raster stats worker result.
2021-07-09 10:56:41,288 (pygeoprocessing.geoprocessing_core) threading.run(870) WARNING No valid pixels were received, sending None.
2021-07-09 10:56:41,289 (taskgraph.Task) Task.add_task(731) ERROR Something went wrong when adding task carbon_map_c_above_cur (0), terminating taskgraph.
Traceback (most recent call last):
  File "natcap\invest\utils.py", line 863, in reclassify_raster
  File "pygeoprocessing\geoprocessing.py", line 1856, in reclassify_raster
  File "pygeoprocessing\geoprocessing.py", line 440, in raster_calculator
  File "pygeoprocessing\geoprocessing.py", line 1849, in _map_dataset_to_value_op
pygeoprocessing.geoprocessing.ReclassificationMissingValuesError: ('The following 11 raster values [52 53 54 55 56 57 58 59 60 61 62] from "C:\\Users\\dmf\\projects\\invest\\data\\bad-data\\Carbon\\lulc_current_willamette.tif" do not have corresponding entries in the ``value_map``: {1: 1.35, 2: 0.45, 3: 0.18, 4: 0.0, 5: 0.9, 6: 0.0, 7: 0.0, 8: 0.0, 9: 0.0, 10: 0.0, 11: 0.0, 16: 0.0, 102: 0.0}.', array([52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62], dtype=int16))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "taskgraph\Task.py", line 697, in add_task
  File "taskgraph\Task.py", line 1174, in _call
  File "natcap\invest\carbon.py", line 442, in _generate_carbon_map
  File "natcap\invest\utils.py", line 871, in reclassify_raster
ValueError: Values in the LULC raster were found that are not represented under the 'lucode' column of the Carbon Pools table. The missing values found in the LULC raster but not the table are: [52 53 54 55 56 57 58 59 60 61 62].
2021-07-09 10:56:41,291 (natcap.invest.utils) utils.prepare_workspace(130) INFO Elapsed time: 0.16s
```

# Checklist
- [ ] Updated HISTORY.rst (if these changes are user-facing)

- [ ] Updated the user's guide (if needed)

### CODE REVIEW

**Comment by phargogh** [comment]
@davemfish Those changing tracebacks are really interesting, so I started poking around the pyinstaller source to see if there was anything informative there.  While I didn't find a satisfying solution, I did come across [this line in the bootloader function](https://github.com/pyinstaller/pyinstaller/blob/e4be1b5c0b15bc3dbf5efec41d315845ea372f64/bootloader/src/pyi_launch.c#L416) that assembles the traceback itself.  Maybe using [`--debug=noarchive`](https://pyinstaller.readthedocs.io/en/stable/usage.html?highlight=--noarchive#how-to-generate) for building the application would change the traceback here?  This is unsatisfying to me because there's no reason the model's `execute` should be eliminated but `utils` kept in the traceback, both are pure-python files.

File: src/natcap/invest/cli.py [hunk 1]
```diff
@@ -587,7 +587,13 @@ def main(user_args=None):
...
             # We're deliberately not validating here because the user
             # can just call ``invest validate <datastack>`` to validate.
-            getattr(model_module, 'execute')(parsed_datastack.args)
+            try:
+                getattr(model_module, 'execute')(parsed_datastack.args)
+            # Graceful exit from this program, otherwise the pyinstaller-made
+            # executeable will exit with 'Failed to execute script cli' on an
+            # uncaught exception. And that only adds unhelpful noise to stderr.
+            except Exception as error:
+                parser.exit(DEFAULT_EXIT_CODE, str(error))
```
Comment (phargogh) [inline src/natcap/invest/cli.py:596]: Wouldn't this also capture and silence the exception itself?  Would it be useful to add `LOGGER.exception(...)`?

**Review by phargogh — CHANGES_REQUESTED** [review]
Makes sense to me and is consistent with how other exceptions are handled in the CLI!  I had one suggestion about logging the exception traceback, but looks great to me otherwise.

File: src/natcap/invest/cli.py (hunk 1 above)
Comment (davemfish) [reply src/natcap/invest/cli.py:596]: Hmm the exception still ends up in the logfile, but now that you mention it, I'm not sure why.

Comment (davemfish) [reply src/natcap/invest/cli.py]: Oh okay it looks like you're exactly right and this explains the difference in the tracebacks, where everything missing from the **after** example  up top is the traceback that's being captured here.

**Comment by davemfish** [comment]
@phargogh thanks for taking a look. Just pushed a change with the missing `LOGGER.exception` call and that yields a traceback that is exactly the same as the **before** example up top. So that's good. 

I guess I was thrown off by the other exceptions that were still being logged - and I'm not really sure how those ended up in the log since all exceptions were being caught, but I guess that's not so related to this issue.

**Comment by phargogh** [comment]
> I guess I was thrown off by the other exceptions that were still being logged - and I'm not really sure how those ended up in the log since all exceptions were being caught, but I guess that's not so related to this issue.

Yeah, that behavior is a little weird ... since the exception doesn't seem to be logged anywhere else, I'm not sure how that would end up in the logfile.  I _would_ expect the exception to go to stdout, though (before this change), so maybe that's how it was ending up in the workbench, via that stream?  In any case, it'd be a separate issue.

**Review by phargogh — APPROVED** [review]

**Pull request merged**